src/util/Taggable.h

```diff
@@ -0,0 +1,101 @@
+#ifndef RIPPLE_UTIL_TAGGABLE_H
+#define RIPPLE_UTIL_TAGGABLE_H
+
+#include <ostream>
+#include <string>
+
+namespace util {
+namespace detail {
+
+/**
+ * @brief This strategy uses an `atomic_uint64_t` to remain lock free.
+ */
+struct UIntTagStrategy
+{
+    static std::string
+    next();
+};
+
+/**
+ * @brief This strategy uses `boost::uuids::uuid` with a static random generator
+ * and a mutex
+ */
+struct UUIDTagStrategy
+{
+    static std::string
+    next();
+};
+
+}  // namespace detail
+
+/**
+ * @brief A base class that allows attaching a tag to a subclass.
+ * @tparam Strategy Specifies how the tag is generated
+ */
+template <typename Strategy = detail::UUIDTagStrategy>
+class Taggable
+{
+private:
+    /**
+     * @brief A simple wrapper for generating and storing tags in a
+     * thread-safe manner.
+     */
+    class Tag final
+    {
+        std::string tag_ = Strategy::next();
+
+    public:
+        /**
+         * @return std::string const& The string representation of the tag
+         */
+        std::string const&
```

Comment:
I am wondering whether we should use a templated type here. In the case where taggable is `atomic_uint64_t`, we are using extra memory by storing a string. 8 bytes for atomic versus 32 bytes for string. Not sure how many of these `Taggable`s will be alive in RAM at once.